cogs5e/models/sheet/coinpurse.py

```diff
@@ -0,0 +1,180 @@
+
+        if coin_type == 'compact':
+            coin_value = self.total
+            coin_type = 'gp'
+        else:
+            coin_value = getattr(self, coin_type)
+        return f"{CoinTypes[coin_type]['icon']} {coin_type}: {coin_value:,}"
+
+    @property
+    def total(self):
+        total = self.gp
+        total += self.pp * 10
+        total += self.ep * 0.5
+        total += self.sp * 0.1
+        total += self.cp * 0.01
+        return total
+
+    @classmethod
+    def from_dict(cls, d):
+        return cls(**d)
+
+    def to_dict(self):
+        return {
+            "pp": self.pp, "gp": self.gp, "ep": self.ep, "sp": self.sp, "cp": self.cp
+        }
+
+    def auto_convert(self, pp: int = 0, gp: int = 0, ep: int = 0, sp: int = 0, cp: int = 0):
+        temp_pp = self.pp
+        temp_gp = self.gp
+        temp_ep = self.ep
+        temp_sp = self.sp
+        temp_cp = self.cp
+
+        pp_change = pp
+        gp_change = gp
+        ep_change = ep
+        sp_change = sp
+        cp_change = cp
+
+        if temp_cp + cp_change < 0:
+            sp_change = ((cp_change + temp_cp) // 10)
+            cp_change += (-1 * sp_change * 10)
+        if temp_sp + sp_change < 0:
+            ep_change = ((sp_change + temp_sp) // 5)
+            sp_change += (-1 * ep_change * 5)
+            temp_sp += (ep_change * 5)
+        if temp_ep + ep_change < 0:
+            gp_change = ((ep_change + temp_ep) // 2)
+            ep_change += (-1 * gp_change * 2)
+            temp_ep += (gp_change * 2)
+        if temp_gp + gp_change < 0:
+            pp_change = ((gp_change + temp_gp) // 10)
+            gp_change += (-1 * pp_change * 10)
+        if temp_pp + pp_change < 0:
+            raise InvalidArgument("You do not have enough coins to cover this transaction.")
+        print(pp_change, gp_change, ep_change, sp_change, cp_change)
+        return pp_change, gp_change, ep_change, sp_change, cp_change
+
+    async def update_currency(self, pp: int = 0, gp: int = 0, ep: int = 0, sp: int = 0, cp: int = 0,
+                              explicit: bool = False, ctx=None):
+        if not all((
+            isinstance(pp, int),
+            isinstance(gp, int),
+            isinstance(ep, int),
+            isinstance(sp, int),
+            isinstance(cp, int)
+        )):
+            raise TypeError("All values must be numeric.")
+
+        if not all((
+            self.pp + pp >= 0,
+            self.gp + gp >= 0,
+            self.ep + ep >= 0,
+            self.sp + sp >= 0,
+            self.cp + cp >= 0
+        )):
+            if explicit and not await confirm(ctx, "You don't have enough of the chosen coins to complete this transaction. "
+                                                   "Auto convert from larger coins? (Reply with yes/no)"):
+                raise InvalidArgument("You cannot put a currency into negative numbers.")
+            pp, gp, ep, sp, cp = self.auto_convert(pp, gp, ep, sp, cp)
```

Comment:
@Croebh 
IMO the `update_currency` method should be a primitive (i.e. synchronous, no side effects like confirming and autoconverting); rather than asking the user for confirmation and doing the conversion using internal variables here, I think we should pull that logic out:
1. Modify `CoinsArgs` to include a `strict` or `explicit` attribute (replacing the tuple that `parse_coins_args` returns)
2. Create a new method to resolve a strict CoinsArgs to a non-strict CoinsArgs that is a valid transaction on a given Coinpurse
3. Confirm with the user there